### PR TITLE
Fix for issue #1066

### LINKED ISSUE #836: Cannot maximize window completely
# Preliminary information
If I drag the window to the top of my screen to "snap" it to full screen or press the maximize button, it will, for a second, maximize, then it goes back to being slightly smaller than my actual screen size.

## Category
BUG

## Bug: Release or Development?
Beta V2, but I'm pretty sure it's been happening for a while on various versions.

# Report
## Current Behavior
See Preliminary Information

## Expected/Wanted/Requested Behavior
Window should snap to full screen

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1. Attempt to maximize the screen
2. ???
3. Not profit.

## OS used:
Windows

## System specifications
i3-3220, GTX 1060 6GB, 8GB RAM, 1920x1080 24" AOC monitor.

## Network
Not relevant

### LINKED ISSUE #880: Chat window scalling issue
# Preliminary information

@DooblyNoobly 's chat window scalling messes with the scale of different sprites in the game window.

## Category
BUG

## Bug: Release or Development?
Development

# Report
## Current Behavior
image
After testing on the inbuilt unity player and the steam build, the problem seems to persist, and that's squished pixels either vertically or horizontally on multiple lines of tiles, and anything over them, so it ends up being more or less than 64x64 (the expected size).

## Expected/Wanted/Requested Behavior
image
Locking the aspect ratio to 1280x720 and zooming in by 1x has fixed the problem on the editor, and painfully resizing the chat to the perfect distance to not have squished pixels did it on the build.
But anyway, the wanted result is all tiles being locked to an even number of pixels (64x64 on bigger monitors, and 32x32 on smaller ones), anything inbetween will cause this, the picture above is pretty much how it should look.

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1.Open the current build or the unity game window.
2.Move around your character and see how some pixels can get squished either horizontally or vertically.
image

## OS used:
Windows

## System specifications
1366x768 monitor.

## Network
Irrelevant

### LINKED ISSUE #1066: 1-8x Game scaling setting (pixel perfect), unlock aspect ratio
# Preliminary information

## Category
Feature Request

# Report
## Current Behavior
Camera size setting is fixed to 5.6666, aspect ratio is locked 16:9

## Expected/Wanted/Requested Behavior
Add a setting that would allow changing camera zoom from 1x to 8x.
Camera size setting should be recalculated automatically (on game window resize) using following pixel perfect formula: 
 ( (height/width)*10 ) * 2 / zoom_amount
So for 16:9 ratio and 2x scaling it would equal 5.625 .
Aspect ratio enforcing should be turned off.

This would likely fix #836 , #880 and maybe #1037

### PR DESCRIPTION
**Pull request by NinjaDanz3r**
### Purpose
This PR adds pixel perfect zooming and resolution. 
Fixes #1066 and fixes #880. 
#1037 and #836 will probably be fixed by some additional taming of ResponsiveUI.
Obligatory gif:
https://imgur.com/a/WdmHUcX

### Approach
I worked together with @fomalsd in order to resolve this issue, with some additional background gained through my attempt to fix #836. See that issue for some additional reading.

### Open Questions and Pre-Merge TODOs

- [x]  I read the contribution guidelines and the wiki.
- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code.
- [x]  This PR does not include files without the specific need to do so.
- [x]  This PR does not bring up any new compile errors.
- [x]  This PR has been tested in the editor.
- [x]  This PR has been tested in a standalone build.
- [x]  This PR has been tested in fullscreen mode.
- [ ]  This PR has been tested in multiplayer.

### Notes:
The camera might not seem to center on the player, but that is due to the camera being offset by the chat window.

### CODE REVIEW

**Comment by DooblyNoobly** [comment]
When the camera size changes then the camera needs to be recentered. There is a function somewhere that does it when the right panel is resized. See if you can find it and call that function straight after the resizing.

**Comment by fomalsd** [comment]
Yeah, and also consider changing zoom positions to 1,2,3,4,5,8 or something like that, no way anyone's going to use 16x, while 3x is native zoom for 1080p

**Comment by fomalsd** [comment]
Also did you check if/how it works in fullscreen? That ResponsiveUI was doing nasty things in fullscreen

**Comment by NinjaDanz3r** [comment]
@fomalsd Latest commit tested in fullscreen on a standalone build, no noticeable errors on my end.

File: UnityProject/Assets/Scripts/UI/CameraZoomHandler.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		SetZoomLevel(-1);
+	}
+	
+	// Update is called once per frame
+	void Update()
+	{
+		if(cachedZoomLevel != zoomLevel)
+		{
+			// Calculate ratio.
+			double ratio = Screen.height / (double) Screen.width;
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/UI/CameraZoomHandler.cs]: Hm, what's the point of placing it in Update() if it's only getting changed via SetZoomLevel() anyway?
Why don't you convert this to Refresh() method that would recalculate camera size based on updated window size and just use this.zoomLevel ?

File: UnityProject/Assets/Scripts/UI/CameraZoomHandler.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using UnityEngine;
+
+public class CameraZoomHandler : MonoBehaviour
+{
+	private int zoomLevel = 1;
+	private int cachedZoomLevel = 1;
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/UI/CameraZoomHandler.cs]: I don't think you'll need this after all, read below

File: UnityProject/Assets/Scripts/UI/CameraZoomHandler.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+			double scaleFactor = Math.Sqrt(Screen.height * Screen.width / (409600 * ratio));
+			// Calculate orthographic size with full precision and then convert to float precision.
+			Camera.main.orthographicSize = Convert.ToSingle(ratio * 10 * scaleFactor / zoomLevel);
+			// Recenter camera.
+			DisplayManager.Instance.SetCameraFollowPos();
+			cachedZoomLevel = zoomLevel;
+		}
+	}
+
+	public void SetZoomLevel(float zoomLevel)
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/UI/CameraZoomHandler.cs]: Looks like this never gets triggered on resolution change, why not make this method only change this.zoomLevel, and Refresh() afterwards?

**Review by fomalsd — CHANGES_REQUESTED** [review]
Camera size should be updated every time window size is changed, currently it only happens on the setting change

File: UnityProject/Assets/Scripts/UI/CameraZoomHandler.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        public void SetZoomLevel(float zoomLevel)
+        {
+            int integerZoomLevel = (int)zoomLevel;
+		    
+            // Automatically try to determine zoom level if we're feeding it some dumb values, such as when this component initializes.
+            if (integerZoomLevel < 1 || integerZoomLevel > 16)
+            {
+                double ratio = Screen.height / (double) Screen.width;
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/UI/CameraZoomHandler.cs]: You don't need this logic here anymore :) just keep the last 2 lines of this method, Refresh() should have all the logic including zoom level check and text updates. Other than that, it looks fine!

**Review by fomalsd — APPROVED** [review]
Code is wonderful! 
However when checking it out I found that I still can't maximize window to fullscreen – ResponsiveUI resets it to windowed mode, so it does not fix #836 and #1037 at the moment. 
Still going to merge it as it gives pixel perfect results at any custom resolution in builds, well done!

**Pull request merged**